### PR TITLE
openjdk-source: new port

### PR DESCRIPTION
**Pull request by usersxx**
#### Description
A port for building openjdk

###### Type(s)

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on

macOS 10.13.6 17G65 x86_64
XCode Command Line Tools Only

###### Verification 
Have you

- [ ] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [x] checked your Portfile with `port lint --nitpick`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by usersxx** [comment]
@herbygillot could you check the portfile.
`openjdk8u` fails on macos-11 workflows during `build changed subports` phase because xcode 13 is used while `openjdk8u` builds from xcode 9 to xcode 12.
If `openjdk13u` and `openjdk15u` fails on macos-11 workflows during `build changed subports` phase, it is due to a bug.
Due to above reasons, macos-10.15 workflow might cancel but if `openjdk13u` passes on macos-10.15 workflows during `build changed subports` phase, `openjdk15u` might build properly on macos 10.15.

**Comment by usersxx** [comment]
@herbygillot could you check the Portfile and after check if everything is ok, you may merge this pull request. Before checking Portfile you should first read the points given below about workflow runs

- `openjdk8u` fails on macos-11 workflows during `build changed subports` phase because xcode 13 is used while `openjdk8u` builds from xcode 9 to xcode 12.
- If `openjdk13u` and `openjdk15u` fails on macos-11 workflows during `build changed subports` phase, it is due to a bug.
- Due to above reasons, macos-10.15 workflow might cancel but if `openjdk13u` passes on macos-10.15 workflows during `build changed subports` phase, `openjdk15u` might build properly on macos 10.15.

**Comment by usersxx** [comment]
@herbygillot Could you merge this pr

**Comment by usersxx** [comment]
@herbygillot you may merge this pr

**Comment by usersxx** [comment]
@l2dy you may merge this pull request

**Comment by usersxx** [comment]
@mascguy you may merge this pull request after checking the Portfile

**Comment by herbygillot** [comment]
Hey @usersxx.  Thanks for your submission.

I think the best person to comment would be a maintainer of current OpenJDK ports, such as @breun. I'm not an expert on the Java ecosystem, though I can help comment on some particulars.

I guess the first question I have is, we already have an OpenJDK Portfile.

Why this new one?

**Comment by usersxx** [comment]
> I guess the first question I have is, we already have an OpenJDK Portfile.
> 
> Why this new one?

This is a different port as it builds OpenJDK from sources instead of installing binaries from any vendor

**Comment by usersxx** [comment]
Though the boot-jdk provided here are from different vendor

File: java/openjdk-java/Portfile
Comment (herbygillot) [inline java/openjdk-java/Portfile:140]: Do not call `sysctl` directly.  MacPorts already provides a variable for this: `${build.jobs}`:

```suggestion
    build.post_args     JOBS=${build.jobs} \
```
```

**Comment by mascguy** [comment]
> I think the best person to comment would be a maintainer of current OpenJDK ports, such as @breun. I'm not an expert on the Java ecosystem, though I can help comment on some particulars.

Seems reasonable. I'm also happy to provide guidance and recommendations though, if folks are open to it. (Nearly 20 years of experience developing for Java here.)

File: java/openjdk-java/Portfile
Comment (usersxx) [reply java/openjdk-java/Portfile:140]: I have set to `sysctl -n hw.ncpu` instead of `${build.jobs}` because parralel build is disabled so `build.job = 1` and if I enable `parralel_build`, it would pass `-j${build.jobs}` to make which isn't supported in the OpenJDK repository and will give error.

Comment (usersxx) [reply java/openjdk-java/Portfile:140]: Though I can remove `sysctl -n hw.ncpu` though as during configure, jobs is already set based on Random-access memory

**Comment by herbygillot** [comment]
Another important point going forward is that there are guidelines on the commit for changes. These are mentioned as [Submission Guidelines](https://trac.macports.org/wiki/CommitMessages) in the checklist.

1. Commit messages should look like: `<port name>: message...`, so for example the commit message for this submission might look like: `openjdk-java: new port`
2. Changes should be kept in a single commit.  As you make corrections and edits from feedback, please squash them into the same single commit with the commit message being as described above.

File: java/openjdk-java/Portfile
Comment (herbygillot) [inline java/openjdk-java/Portfile:56]: For `description` fields, if using another variable, please use Tcl's argument expansion (`{*}`):
```suggestion
    long_description    {*}${long_description_adoption-jdk}
```

**Comment by usersxx** [comment]
> Another important point going forward is that there are guidelines on the commit for changes. These are mentioned as [Submission Guidelines](https://trac.macports.org/wiki/CommitMessages) in the checklist.
> 
> 1. Commit messages should look like: `<port name>: message...`, so for example the commit message for this submission might look like: `openjdk-java: new port`
> 2. Changes should be kept in a single commit.  As you make corrections and edits from feedback, please squash them into the same single commit with the commit message being as described above.

I would squash the commits after all changes are made

**Comment by usersxx** [comment]
You may comment if you have suggested all changes, i would squash all commits

File: java/openjdk-java/Portfile
Comment (herbygillot) [inline java/openjdk-java/Portfile:289]: Installing things outside of the prefix is **strongly** frowned upon, in general.

Why is this needed?

File: java/openjdk-java/Portfile
Comment (herbygillot) [inline java/openjdk-java/Portfile:295]: A port should not ever be installing anything outside of the MacPorts prefix.  Please change it so that it installs within the MacPorts tree.

Comment (usersxx) [reply java/openjdk-java/Portfile:295]: I can't actually change that, it is the default path where the file needs to be installed.

Comment (usersxx) [reply java/openjdk-java/Portfile:295]: If the file default installation path was `/usr/local`, I would have changed it to `${prefix}` because /opt/local is a system directory like /usr/local but this files installation directory is `/Library/Java/JavaVirtualMachine`

File: java/openjdk-java/Portfile
Comment (breun) [reply java/openjdk-java/Portfile]: @herbygillot I'm not very familiar with Tcl, and I wonder why this is necessary. It seems this pattern was copied from what I do in the existing `openjdk` portfile.

**Comment by usersxx** [comment]
> > I think the best person to comment would be a maintainer of current OpenJDK ports, such as @breun. I'm not an expert on the Java ecosystem, though I can help comment on some particulars.
> 
> Seems reasonable. I'm also happy to provide guidance and recommendations though, if folks are open to it. (Nearly 20 years of experience developing for Java here.)

@mascguy Is there something, I can change in this port or subports

File: java/openjdk-java/Portfile
Comment (usersxx) [reply java/openjdk-java/Portfile]: `{*}` is required for `description` field, if using another variable, as it promotes the text in the variable to individual arguments.

**Comment by mascguy** [comment]
> @mascguy Is there something, I can change in this port or subports

I'll try to take a look soon, over the next day or two.

Presently, my only question relates to port naming. For example, should we designate this differently, since it's building from source? Etc.

But I'll also take a closer look at the Portfile as well, per the comments from Herby and others.

File: java/openjdk-java/Portfile
Comment (herbygillot) [reply java/openjdk-java/Portfile]: That's correct @usersxx.  @breun `{*}` is Tcl's argument expansion operator.  It takes a group containing multiple items and expands the items out to be individual values.

```
% p {a b c}
Number of arguments p sees: 1
% p {*}{a b c}
Number of arguments p sees: 3
```

...from https://wiki.tcl-lang.org/page/%7B%2A%7D

**Comment by usersxx** [comment]
Though the port and its subports install everything correctly, except `openjdk8u` on macos-11 because macos-11 uses xcode 13 while this port builds on xcode versions before xcode 12, except `openjdk13u` and `openjdk15u` on macos-11 due to a bug in the source and due to error in macos-11 workflows, macos-10.15 workflows get cancelled.

**Comment by usersxx** [comment]
> Presently, my only question relates to port naming. For example, should we designate this differently, since it's building from source? Etc.

I actually don't want to change the name as it builds from repository that is maintained by OpenJDK community and these Subports build OpenJDK without any modification to any component Unlike any vendors would do.

**Comment by usersxx** [comment]
@ryandesign Can you check the portfile

**Comment by usersxx** [comment]
@herbygillot Is it possible to use `PortGroup           github 1.0` to use `github.setup`

**Comment by usersxx** [comment]
> > @mascguy Is there something, I can change in this port or subports
> 
> I'll try to take a look soon, over the next day or two.
> 
> Presently, my only question relates to port naming. For example, should we designate this differently, since it's building from source? Etc.
> 
> But I'll also take a closer look at the Portfile as well, per the comments from Herby and others.

@mascguy Is there anything to change ?

**Comment by herbygillot** [comment]
@usersxx To answer your previous question, yes, you can declare and use the `github` portgroup within subports.  No restriction there.

Another thing I note is that this submission makes a lot of assumptions about the presence of system-prefixed JDKs.  I anticipate it's going to get hairy building across the various macOS versions.

**Comment by mascguy** [comment]
> @mascguy Is there anything to change ?

While I haven't tested this port, the portfile itself looks very clean and easy to understand. And no concerns overall.

Otherwise, I'll defer to Herby.

**Comment by mascguy** [comment]
I tried installing the various subports, and I'm seeing build failures for several of the non-boot variations. These errors occur when the build starts, and at least some appear to be configure-related failures.

I can provide details if needed, but just a proactive heads-up.

**Comment by usersxx** [comment]
> I tried installing the various subports, and I'm seeing build failures for several of the non-boot variations. These errors occur when the build starts, and at least some appear to be configure-related failures.
> 
> I can provide details if needed, but just a proactive heads-up.
> 
> p.s. I'm testing on macOS 10.13.

I had failures with `openjdk8u` (xcode not installed) and `openjdk17u` while configuring

**Comment by usersxx** [comment]
> I tried installing the various subports, and I'm seeing build failures for several of the non-boot variations. These errors occur when the build starts, and at least some appear to be configure-related failures.
> 
> I can provide details if needed, but just a proactive heads-up.
> 
> p.s. I'm testing on macOS 10.13.

I had tested building the repositories manually and not by ports. I have no idea about it, I tried `sudo port -D /path/to/directory/in/which/portfile/exists/ -vst install $subport`

**Comment by usersxx** [comment]
> > I tried installing the various subports, and I'm seeing build failures for several of the non-boot variations. These errors occur when the build starts, and at least some appear to be configure-related failures.
> > I can provide details if needed, but just a proactive heads-up.
> > p.s. I'm testing on macOS 10.13.
> 
> I had tested building the repositories manually and not by ports. I have no idea about it, I tried `sudo port -D /path/to/directory/in/which/portfile/exists/ -vst install $subport`

I didn't use root directory

**Comment by usersxx** [comment]
> I tried installing the various subports, and I'm seeing build failures for several of the non-boot variations. These errors occur when the build starts, and at least some appear to be configure-related failures.
> 
> I can provide details if needed, but just a proactive heads-up.
> 
> p.s. I'm testing on macOS 10.13.

I have figured it out.

**Comment by breun** [comment]
This portfile is called `openjdk-java`, while there is already a portfile called `openjdk`, which contains all the vendor-provided JDK's. This might be a bit confusing.

I know that both `openjdk` and `openjdk-java` are both actually not installable ports, and they only exist to allow grouping a bunch of subports with common functionality.

It might be cleanest if the subports in this new submission would be added to the existing `openjdk` portfile, although there is the issue of maintainer responsibility, because while I'm fine with being the maintainer for the current ports in the `openjdk` portfile, I'm not ready to maintain these from source builds.

Another route might be separating out the JDK's to separate portfiles per vendor (including this one for 'no vendor'), since the `openjdk` portfile is already rather large.

**Comment by herbygillot** [comment]
I think I agree that these from-source JDKs might be better in their own Portfiles.  They don't seem to share much, so I don't see what the benefit is having them as subports within the same Portfile.  Especially since the build process is so involved for each JDK, it will be cleaner understanding when things fail across the buildbots per macOS version.

**Comment by breun** [comment]
I'm also fine with keeping them in separate files. It's just that `openjdk` and `openjdk-java` seem like confusing names in that scenario, because they're both about Java, so the names don't really tell what the differences are.

I'd also be fine with renaming the current `openjdk` portfile to something like `openjdk-distributions`, and then this new submission could be called `openjdk`, or `openjdk-source` or something more descriptive like that.

**Comment by usersxx** [comment]
> This portfile is called `openjdk-java`, while there is already a portfile called `openjdk`, which contains all the vendor-provided JDK's. This might be a bit confusing.
> 
> I know that both `openjdk` and `openjdk-java` are both actually not installable ports, and they only exist to allow grouping a bunch of subports with common functionality.
> 
> It might be cleanest if the subports in this new submission would be added to the existing `openjdk` portfile, although there is the issue of maintainer responsibility, because while I'm fine with being the maintainer for the current ports in the `openjdk` portfile, I'm not ready to maintain these new builds from source.
> 
> Another route might be separating out the JDK's to separate portfiles per vendor (`openjdk-corretto`, `openjdk-zulu`, etc.), since the `openjdk` portfile is already rather large. And then maybe these builds from source could be in the `openjdk` portfile in the end.

This port is set to `nomaintainer` and thus with no maintainer. So you don't need to maintain. I agree that this portfile should split so there will be each portfile for each vendor because of the portfile being too large. (future LTS releases with azul zulu MTS releases)

> I'm also fine with keeping them in separate files. It's just that `openjdk` and `openjdk-java` seem like confusing names in that scenario, because they're both about Java, so the names don't really tell what the differences are.
> 
> I'd also be fine with renaming the current `openjdk` portfile to something like `openjdk-distributions`, and then this new submission could be called `openjdk`, or `openjdk-source` or something more descriptive like that.

If you don't want to split `openjdk` port into different ports, I agree that openjdk meta port name should change from `openjdk` to `openjdk-distributions` while this openjdk-java port should change from `openjdk-java` to `openjdk-source` to show that `openjdk-distributions` distributes jdk while `openjdk-source` builds from source.

**Comment by usersxx** [comment]
> I tried installing the various subports, and I'm seeing build failures for several of the non-boot variations. These errors occur when the build starts, and at least some appear to be configure-related failures.
> 
> I can provide details if needed, but just a proactive heads-up.
> 
> p.s. I'm testing on macOS 10.13.

@mascguy I have successfully installed `openjdk11u`, `openjdk13u` and `openjdk15u` but `openjdk8u` failed because xcode wasn't installed and `openjdk17u` failed because metal wasn't found (I have a old mac that doesn't support metal requirements)

**Comment by breun** [comment]
I've created https://github.com/macports/macports-ports/pull/14212 to rename the `openjdk` port to `openjdk-distributions` to allow making a clear distinction between the subports in that portfile and the ones in this new proposed one.

**Comment by usersxx** [comment]
> I've created #14212 to rename the `openjdk` port to `openjdk-distributions` to allow making a clear distinction between the subports in that portfile and the ones in this new proposed one.

Good, I assume that #14185 should be merged with #14212

**Comment by usersxx** [comment]
> @mascguy, you can merge this pr

Don't merge it yet

**Comment by mascguy** [comment]
> @mascguy I have successfully installed `openjdk11u`, `openjdk13u` and `openjdk15u` but `openjdk8u` failed because xcode wasn't installed and `openjdk17u` failed because metal wasn't found (I have a old mac that doesn't support metal requirements)

Sounds good.

We'll definitely want to ensure all of these work before merging, though.

I'm happy to help with this, if you need assistance. Let me know!

**Comment by usersxx** [comment]
setting `use_xcode` to yes will not allow to install these ports for users without xcode while setting `use_xcode` to no will not allow this port to use xcode tools like `xcodebuild`

**Comment by usersxx** [comment]
> > @mascguy I have successfully installed `openjdk11u`, `openjdk13u` and `openjdk15u` but `openjdk8u` failed because xcode wasn't installed and `openjdk17u` failed because metal wasn't found (I have a old mac that doesn't support metal requirements)
> 
> Sounds good.
> 
> We'll definitely want to ensure all of these work before merging, though.
> 
> I'm happy to help with this, if you need assistance. Let me know!

Is there a way to check if a directory exists?

**Comment by usersxx** [comment]
> 
> @mascguy I have successfully installed `openjdk11u`, `openjdk13u` and `openjdk15u` but `openjdk8u` failed because xcode wasn't installed and `openjdk17u` failed because metal wasn't found (I have a old mac that doesn't support metal requirements)

Currently with XCode Cl it builds normally and checking workflows, it builds with XCode app properly too. @mascguy, I think this pr can be merged

**Comment by usersxx** [comment]
> > @mascguy I have successfully installed `openjdk11u`, `openjdk13u` and `openjdk15u` but `openjdk8u` failed because xcode wasn't installed and `openjdk17u` failed because metal wasn't found (I have a old mac that doesn't support metal requirements)
> 
> Currently with XCode Cl it builds normally and checking workflows, it builds with XCode app properly too. @mascguy, I think this pr can be merged

Currently `openjdk8u` fails on macos 11 because the port doesn't support xcode 13 while `openjdk13u` and `openjdk15u` fails due to bug in the source code.

**Comment by usersxx** [comment]
@mascguy What do you think?

**Comment by usersxx** [comment]
> > @mascguy I have successfully installed `openjdk11u`, `openjdk13u` and `openjdk15u` but `openjdk8u` failed because xcode wasn't installed and `openjdk17u` failed because metal wasn't found (I have a old mac that doesn't support metal requirements)
> 
> Sounds good.
> 
> We'll definitely want to ensure all of these work before merging, though.
> 
> I'm happy to help with this, if you need assistance. Let me know!

I don't think that there is any thing to change as my mac is doesn't support metal due to which, `openjdk17u` fails and it only builds with metal. Moreover, `openjdk8u` requires full XCode app which is not installed on my system and thus it fails. Not every port or subport can support every macOS hardware and macOS version.

**Comment by usersxx** [comment]
> Moreover, `openjdk8u` requires full XCode app which is not installed on my system and thus it fails. Not every port or subport can support every macOS hardware and macOS version.

XCode App as well as XCode Cl would be installed by most users who use MacPorts as MacPorts requires XCode (both App and Cl) which is mentioned in installing MacPorts page. (https://www.macports.org/install.php)

**Comment by mascguy** [comment]
> @mascguy What do you think?

Let me start by saying that we all appreciate your enthusiasm, and willingness to keep refining this work. That's a great thing, and it's much appreciated!

So overall, my advice is this: As difficult as it is, try to be patient, and not rush the effort.

Particularly when a port has this many subports - with significant nuances among them all - it's understandably going to take time to get everything right.

And folks have personal and professional commitments - in addition to other work in the MacPorts space - that sometimes makes it difficult for us to quickly respond.

So I'll get back to this over the coming days. But again, try to be patient!

**Comment by usersxx** [comment]
> > @mascguy What do you think?
> 
> Let me start by saying that we all appreciate your enthusiasm, and willingness to keep refining this work. That's a great thing, and it's much appreciated!
> 
> So overall, my advice is this: As difficult as it is, try to be patient, and not rush the effort.
> 
> Particularly when a port has this many subports - with significant nuances among them all - it's understandably going to take time to get everything right.
> 
> And folks have personal and professional commitments - in addition to other work in the MacPorts space - that sometimes makes it difficult for us to quickly respond.
> 
> So I'll get back to this over the coming days. But again, try to be patient!

Ok, I would check this port till then.

**Comment by usersxx** [comment]
@ryandesign can you check the portfile

**Comment by mascguy** [comment]
Folks, while I plan to retest all of these subports when I can find time, it might have to wait until the weekend.

In the interim, if there's anyone else that can find some time to test as well - and/or review the Portfile - that would be awesome!

**Comment by usersxx** [comment]
> Folks, while I plan to retest all of these subports when I can find time, it might have to wait until the weekend.
> 
> In the interim, if there's anyone else that can find some time to test as well - and/or review the Portfile - that would be awesome!

I have checked all the subports.
`openjdk8u` needs xcode and `openjdk17u` needs metal (My mac doesn't support metal)

**Comment by usersxx** [comment]
@mascguy I have currently figured out to install all ports correctly. Only `openjdk8u` fails due to use of newer xcode (xcode 13.2.1) on macos-11 workflow. `openjdk13u` builds properly when creating a server jvm and client jvm combined.  `openjdk15u` is fixed when creating a server jvm and client jvm combined as well as using a [patch](https://raw.githubusercontent.com/usersxx/macports-ports/java--openjdk-java/java/openjdk-source/files/patch-openjdk15u-build-fix.diff). Currently `openjdk17u` requires Xcode for Metal and setting `use_xcode           yes` fixed it

**Comment by usersxx** [comment]
> @mascguy I have currently figured out to install all ports correctly. Only `openjdk8u` fails due to use of newer xcode (xcode 13.2.1) on macos-11 workflow. `openjdk13u` builds properly when creating a server jvm and client jvm combined. `openjdk15u` is fixed when creating a server jvm and client jvm combined as well as using a [patch](https://raw.githubusercontent.com/usersxx/macports-ports/java--openjdk-java/java/openjdk-source/files/patch-openjdk15u-build-fix.diff). Currently `openjdk17u` requires Xcode for Metal and setting `use_xcode yes` fixed it

I though if there are multiple xcode installed is it possible to set maximum xcode version. That would be great.

**Comment by usersxx** [comment]
> @mascguy I have currently figured out to install all ports correctly. Only `openjdk8u` fails due to use of newer xcode (xcode 13.2.1) on macos-11 workflow. `openjdk13u` builds properly when creating a server jvm and client jvm combined. `openjdk15u` is fixed when creating a server jvm and client jvm combined as well as using a [patch](https://raw.githubusercontent.com/usersxx/macports-ports/java--openjdk-java/java/openjdk-source/files/patch-openjdk15u-build-fix.diff). Currently `openjdk17u` requires Xcode for Metal and setting `use_xcode yes` fixed it

I thought is it possible to set the maximum xcode version since a person can install multiple Xcode versions.

**Comment by usersxx** [comment]
> @mascguy I have currently figured out to install all ports correctly. Only `openjdk8u` fails due to use of newer xcode (xcode 13.2.1) on macos-11 workflow. `openjdk13u` builds properly when creating a server jvm and client jvm combined. `openjdk15u` is fixed when creating a server jvm and client jvm combined as well as using a [patch](https://raw.githubusercontent.com/usersxx/macports-ports/java--openjdk-java/java/openjdk-source/files/patch-openjdk15u-build-fix.diff). Currently `openjdk17u` requires Xcode for Metal and setting `use_xcode yes` fixed it

Without creating a server jvm and client jvm combined in `openjdk13u`, it builds correctly. Only `configure.args      --disable-warnings-as-errors` was required.

**Comment by usersxx** [comment]
> > @mascguy I have currently figured out to install all ports correctly. Only `openjdk8u` fails due to use of newer xcode (xcode 13.2.1) on macos-11 workflow. `openjdk13u` builds properly when creating a server jvm and client jvm combined. `openjdk15u` is fixed when creating a server jvm and client jvm combined as well as using a [patch](https://raw.githubusercontent.com/usersxx/macports-ports/java--openjdk-java/java/openjdk-source/files/patch-openjdk15u-build-fix.diff). Currently `openjdk17u` requires Xcode for Metal and setting `use_xcode yes` fixed it
> 
> I thought is it possible to set the maximum xcode version since a user can install multiple Xcode versions.

After all changes done, I would merge the commites

**Comment by usersxx** [comment]
Currently all Subports work fine.

**Comment by usersxx** [comment]
> > @mascguy I have currently figured out to install all ports correctly. Only `openjdk8u` fails due to use of newer xcode (xcode 13.2.1) on macos-11 workflow. `openjdk13u` builds properly when creating a server jvm and client jvm combined. `openjdk15u` is fixed when creating a server jvm and client jvm combined as well as using a [patch](https://raw.githubusercontent.com/usersxx/macports-ports/java--openjdk-java/java/openjdk-source/files/patch-openjdk15u-build-fix.diff). Currently `openjdk17u` requires Xcode for Metal and setting `use_xcode yes` fixed it
> 
> I thought is it possible to set the maximum xcode version since a user can install multiple Xcode versions.

Instead of setting maximum xcode version, I have added a patch that will make `openjdk8u` support xcode 13

**Comment by usersxx** [comment]
> > > @mascguy I have currently figured out to install all ports correctly. Only `openjdk8u` fails due to use of newer xcode (xcode 13.2.1) on macos-11 workflow. `openjdk13u` builds properly when creating a server jvm and client jvm combined. `openjdk15u` is fixed when creating a server jvm and client jvm combined as well as using a [patch](https://raw.githubusercontent.com/usersxx/macports-ports/java--openjdk-java/java/openjdk-source/files/patch-openjdk15u-build-fix.diff). Currently `openjdk17u` requires Xcode for Metal and setting `use_xcode yes` fixed it
> > 
> > 
> > I thought is it possible to set the maximum xcode version since a user can install multiple Xcode versions.
> 
> Instead of setting maximum xcode version, I have added a patch that will make `openjdk8u` support xcode 13

Current, I think all subports are ready

**Comment by usersxx** [comment]
> @mascguy I have currently figured out to install all ports correctly. Only `openjdk8u` fails due to use of newer xcode (xcode 13.2.1) on macos-11 workflow. `openjdk13u` builds properly when creating a server jvm and client jvm combined. `openjdk15u` is fixed when creating a server jvm and client jvm combined as well as using a [patch](https://raw.githubusercontent.com/usersxx/macports-ports/java--openjdk-java/java/openjdk-source/files/patch-openjdk15u-build-fix.diff). Currently `openjdk17u` requires Xcode for Metal and setting `use_xcode yes` fixed it

I have added patchfiles to `openjdk8u` to support newer version of xcode.

**Comment by usersxx** [comment]
> @mascguy I have currently figured out to install all ports correctly. Only `openjdk8u` fails due to use of newer xcode (xcode 13.2.1) on macos-11 workflow. `openjdk13u` builds properly when creating a server jvm and client jvm combined. `openjdk15u` is fixed when creating a server jvm and client jvm combined as well as using a [patch](https://raw.githubusercontent.com/usersxx/macports-ports/java--openjdk-java/java/openjdk-source/files/patch-openjdk15u-build-fix.diff). Currently `openjdk17u` requires Xcode for Metal and setting `use_xcode yes` fixed it

I would test these ports on my mac. Current Xcode isn't installed and thus, only `openjdk8u` and `openjdk17u` can't be installed.

**Comment by usersxx** [comment]
I have squashed all commits and tested all ports. All the port currently build properly and I don't have anything to change

File: java/openjdk-source/Portfile
Comment (mascguy) [inline java/openjdk-source/Portfile:388]: Nit-pick: Remove trailing whitespace on this line.

**Comment by mascguy** [comment]
I provided similar feedback for Nils' existing `openjdk` port, and it applies here as well:

* Don't obsolete the parent port `openjdk-source`: An active, valid port is not obsolete. Even if it's just a stub!
* Instead, use the following pattern:

```
depends_lib-append      <PORT_YOU_WANT_TO_BE_INSTALLED_BY_DEFAULT>
distfiles
patchfiles
supported_archs         noarch
use_configure           no
build {}
destroot {
    set docdir ${destroot}${prefix}/share/doc/${subport}
    xinstall -d ${docdir}
    system "echo $subport is a stub port > ${docdir}/README"
}
```

**Comment by mascguy** [comment]
One more nit-picky suggestion: Update your commit message to 'openjdk-source: new port'.

**Comment by mascguy** [comment]
Updated my previous comments, so be sure to refresh to pick those up!

File: java/openjdk-source/Portfile
Comment (usersxx) [reply java/openjdk-source/Portfile:388]: Oops, I didn't check Portfile with `port lint  --nitpick` after all commits.

**Comment by breun** [comment]
> * Don't obsolete the parent port `openjdk-source`: An active, valid port is not obsolete. Even if it's just a stub!

I personally don't see a use case for a port that installs the latest version of OpenJDK, and then updates to a possibly incompatible new major version when that comes out, so I'd personally prefer to keep the main port non-installable and only let users install specific major versions.

**Review by mascguy — APPROVED** [review]

**Comment by mascguy** [comment]
> @mascguy I have squashed all commits and tested all ports. All the port currently build properly and I don't have anything to change.

Great work on this!

**Pull request merged**
